src/main/scala/deduplication/DeduplicationContext.scala

```diff
@@ -27,6 +27,20 @@ trait DeduplicationContext[F[_], ID, ContextID, A] {
     * @return the result of [[fa]]
     */
   def protect(id: ID, fa: F[A]): F[A]
+
+  /**
+    * Do a best effort attempt at ensuring a process [[fa]] is successfully executed only once.
+    *
+    * If the process has already ran successfully before, it will return the original result of
+    * [[fa]]
+    * Otherwise, it will run [[fa]].
+    *
+    * @param id The id of the process to run.
+    * @param fa The effect to run if the process is new.
+    * @param onDeduplicateDetected An effect that's run when a duplicate is detected. It can be useful for logging.
+    * @return the result of [[fa]]
+    */
+  def protect(id: ID, fa: F[A], onDeduplicateDetected: A => F[Unit]): F[A]
```

Comment:
I played with the idea of just having a single method with a default value, like this:
```scala
def protect(id: ID, fa: F[A], onDeduplicateDetected: A => F[Unit] = _ => Applicative[F].unit): F[A]
```

but the `trait` doesn't require any typeclasses so it doesn't have an instance for `Applicative[F]`.